### DISCUSSION TITLE
PvE difficulty in E6

### DISCUSSION THREAD

**Discussion by willbattel** [NillerMedDild/Enigmatica6#3216] [General]
Hello all, I want to pose this question to both the developers and players of Enigmatica 6. Some friends and I recently wrapped up a 0.4.X server we started in February and are thinking about our next SMP server.

We really like Enigmatica, but one thing we consistently noticed was how invulnerable the players became with the combination of enchantments, trinkets, [SoL: CE](https://www.curseforge.com/minecraft/mc-mods/spice-of-life-carrot-edition) health bonuses, and other miscellaneous means of buffing the player. Meanwhile, the hostile mobs do not get any more difficult- such that even on "Hard" they become mere nuisances to players still in the early-mid game stages.

We recognize that this is a "kitchen sink" pack, not focused on survival or combat, but I wanted to pose the question of PvE difficulty to the community to see what others felt about PvE difficulty. Are there readily-available server configs that should be adjusted for administrators looking to give players a reason to fear mobs again? Can the bosses be made more difficult? Are there mods are currently included that make it too easy for players? Are there mods available, not currently included, that could make PvE more threatening?

For example, my group would prefer to replace SoL:CE with [the original SoL](https://www.curseforge.com/minecraft/mc-mods/the-spice-of-life), removing the health bonuses (my friend achieved over 100HP this way, and that isn't even the limit when combined with the trinkets) and requiring greater attention to food, but just by comparing their download-count it's easy to see that the community is divided.

My goal is not to make PvE annoyingly difficult- my friends and I just got bored with how OP we were so early on in our play-through. If that's how E6 is meant to be, then that's okay- I just wanted to talk about it.

**Comment by MuteTiefling** [comment]
So, I kinda agree with where you're coming from. But there are some limiting factors at play.

1) Minecraft combat is terrible. Just awful. There's no nuance to it and it wasn't really a system built to do much to begin with. 
2) With enchanted Vanilla gear (diamond, netherite, whatever) you already become  nearly untouchable to mobs. The thing most likely to kill you late game is yourself. Oops, elytra into a wall. Oops, fell into the void. Dang 200 block falls hurt.

Now, factor in mods and the result is you go from nearly unkillable to very much unkillable in short order. Be it from higher level enchants, better armor, etc.

So what does one do? There are a few things that can be done, actually. But they aren't perfect. 
1) Make mobs very powerful
2) Limit access to powerful gear until much later

Very powerful mobs means you're basically screwed at the start and very likely to die frequently. This is exceedingly frustrating for some who just want to build and goof around. This isn't dark souls, after all. The focus isn't combat. It's building. Again, there are a few ways to deal with this, which I'll get into. 

Limiting access to gear is a slightly better approach. We all know we can survive a few mobs in terrible gear. We do it regularly in the early game. It can get a bit tedious though as you're constantly interrupted with protracted fights when you just wanna mine them shiny rocks, darn it! It does have the benefit of making those armor upgrades feel really good to actually craft though. You've suffered a bit without it, then finally you can make good armor and relax a bit!

Back to the very powerful mobs. So let's say we _do_ want really big mean mobs. Well, obviously spending hours dying and reviving gets old pretty quick. So we can scale them, make them rare, or stage them. 

Scaling Mobs: Scaling works by having the mobs weak or normal power level at the start, then ramping up their stats as game time progresses. This might seem OK, but... your player power isn't necessarily growing at a constant rate. So now you feel like you have to rush to powerful gear to stay ahead of the scaling. That's not always fun.

Rare Mobs: This is what we actually have right now with Apotheosis bosses. They spawn rarely and you generally have the choice to engage. They can be rough, but they're easy to run away from too. 

Staged Mobs: This can work well, but it's a lot of work to set up. Essentially you limit what mobs are allowed to spawn via game stages. SevTech did this fairly well. Early on, you'd never encounter a skeleton with a bow, for instance. Only melee skeletons. So the threat to the player was reduced since you didn't have to worry about getting sniped in your crappy gear. Then as you unlock stages, more and more mob types start spawning. You can even start including some hard hitters this way. None of these systems are mutually exclusive, after all. It's a decent system, it just takes a lot of effort to set up and is probably out of scope for a kitchen sink pack. This is the kind of thing you need to build a pack around.

All that aside, going back to MC combat... as stated, there's no nuance. The only way to really make a fight 'hard' is to ramp up health/armor/strength of mobs since their AIs are simplistic. But because of the way the armor system works, the same thing that makes you nearly invincible to mobs quickly makes them nearly unkillable in return. There's this magic spot where things go from being able to one shot you, to just tickling you... and as soon as you hit that spot, combat becomes trivial. There's not a little wiggle room between these states. So you're either stuck in this 'every mob I meet destroys me' state, or you get over the hump, so to speak, and become unkillabe. 

Now, does that mean we can't do anything? No. Will we? Maybe? I can't really say there. I will say that we've made a bit of an attempt at making the early game a bit more of a challenge for Expert, however, by making armor, especially good armor, a little harder to come by. Not a lot, but you'll have to work for it more than normal. You can't just dig to bedrock and make a suit of diamond armor, for instance. You might be wearing iron armor longer than you'd expect. And progression will have you facing bosses fairly early: hopefully earlier than you might think you're ready for, but not so early that they're impossible. You'll have to take on some pharaohs before you've got good enchants or even diamond armor, for instance, and an early trip to the End is likely. Though technically you won't need to kill the dragon then, I'd suspect most will. 

There's also an important point to factor in: Enigmatica series hasn't ever really been focused on combat. It's been more focused on the automation and building aspects of modded, with some exploration sprinkled in to keep you from just sitting at a crafting grid in your base the entire time you're playing. We try to balance things out a bit. Some combat, some building, some exploration. Too much of one thing gets old, after all.

**Reply by willbattel** [reply]
I agree with basically all of your notes here. Minecraft is not a combat-focused game, and Enigmatica is not a combat-focused pack. I don't think there is any expectation of revolutionary changes to the combat system, but I do think that, as a relatively new and quickly-iterating pack, E6 will have the opportunity to tune this behavior over time to improve that balance you point out as difficult.

The scaling method would have race conditions not conducive to an improved experience, and staging is a major undertaking that might not belong in this pack. However, I would argue that rare mobs could be further buffed. My friends and I logged a logged a significant amount of time on our 0.4 server, and none of the bosses we faced posed a threat. We probably didn't face them all, I didn't check, but the variety we encountered were easily defeated or avoided without much thought, which is unfortunate.

So that leaves me with a few points:

1. I think sometime in the future when the pack is more mature, the devs might want to make another balancing pass and closely evaluate the bosses in particular. As you mentioned: the combat mechanics are very limiting, but I think even a few extra HP or damage here and there (maybe even movement speed if you can easily change that) could help.
2. Are there server configurations available to admins that want to toy with mob difficulty beyond the basic Easy/Normal/Hard options? Everyone wants different things, so rather than changing the default behavior, maybe we can get away with what's available in the configs.
3. I think above all else, my biggest gripe is with Spice of Life: Carrot Edition. Having such an easy way to grow your health might be the biggest contributor to the problem. We ran around finding and crafting food for a couple hours and got an enormous health bar that we never had to worry about again. My group in particular also wants the negative consequences dealt by the original SoL mod but, as I said above, everyone wants different things. ~I know that when you use a modpack like this, you're usually signing up for the whole thing as-is, but I'm wondering how challenging it would be for interested server admins to swap Carrot Edition with the original SoL variant- as both variants are popular. I kinda wish they were just one mod that could be configured to behave one way or the other...but since they're not, I'm trying to get creative. Is it possible for E6 to ship with both mods and provision the swapping between them? There isn't really a perfect solution available here... I'm just brainstorming because we really want E6 but with Carrot Edition replaced with the original. We can of course make a modified pack just for our group- but then it becomes difficult for us to keep said fork up to date among the other frictions of pack maintenance and distribution you're familiar with.~

EDIT: Just noticed that the original SoL isn't available for 1.16... so scratch that thought 🙃 but at least we can reduce/nullify the Carrot Edition buffs in the configs.

**Reply by NielsPilgaard** [reply]
I agree that the possible benefits from SoL: CE are far too great, I've just nerfed it to a maximum of 10 bonus hearts, down from 28.

**Reply by NielsPilgaard** [reply]
You can configure this yourself in the server config `solcarrot-server.toml` btw :)

**Reply by dudeloony** [reply]
Just read this discussion, and I think SOL: potato edition may be the way to go. You can read the description here: [https://www.curseforge.com/minecraft/mc-mods/spice-of-life-potato-edition] but it seems to be really well balanced and makes food seeking really relevant. Also, it's highly configurable, so there's that.

**Reply by willbattel** [reply]
I didn't know about potato edition, but having to maintain the diversity to keep the buffs instead of getting them permanently sounds great to me!

**Reply by NielsPilgaard** [reply]
We already have that through Diet though.
